xrtpy/response/xrt_teem.py

```diff
@@ -0,0 +1,428 @@
+    the image size is not reduced, but pixels within 3×3 squares are set to
+    the same value, which results from averaging over those pixels.
+
+    Notes
+    -----
+    The returned values of pixels where the temperature cannot be derived
+    or the error is greater than the threshold or the photon noise is
+    greater than the threshold are set to 0. The EM for those pixels is
+    also set to 0.
+
+    The details of the coronal-temperature-diagnostic capability of
+    Hinode/XRT is described in
+    Narukage et al. 2011, Solar Phys., 269, 169.
+    http://adsabs.harvard.edu/doi/10.1007/s11207-010-9685-2
+    and
+    Narukage et al. 2013, Solar Phys.,
+    http://adsabs.harvard.edu/doi/10.1007/s11207-013-0368-7
+    These two papers are the reference papers of this program.
+
+    Modification History:
+
+    IDL routine written by N.Narukage (NAOJ). Converted to python by Jonathan
+    D. Slavin (SAO). See original IDL code for more details.
+    """
+
+    hdr1 = map1.meta
+    hdr2 = map2.meta
+    data1 = map1.data
+    data2 = map2.data
+    n1 = "XRT_RENORMALIZE" in hdr1["HISTORY"]
+    n2 = "XRT_RENORMALIZE" in hdr2["HISTORY"]
+    # This allows use of normalized data (contrary to original IDL code):
+    if n1 or n2:
+        if n1 and n2:
+            data1 = data1 * hdr1["EXPTIME"]
+            data2 = data2 * hdr2["EXPTIME"]
+        elif n1:
+            data1 = data1 * hdr1["EXPTIME"]
+        else:
+            data2 = data2 * hdr2["EXPTIME"]
+    if data1.shape != data2.shape:
+        raise ValueError("The input images must be the same size")
+
+    if mask is None:
+        mask = np.zeros_like(data1, dtype=bool)
+    data1 = data1.astype(float)
+    data2 = data2.astype(float)
+    data1 = np.ma.masked_where(mask, data1)
+    data2 = np.ma.masked_where(mask, data2)
+    if binfac > 1:
+        s = data1.shape
+        ns = (s[0] // binfac, s[1] // binfac)
+        rbs = (ns[0], binfac, ns[1], binfac)
+        # sums the data in binfac x binfac sized regions
+        d1 = data1.reshape(rbs).sum(-1).sum(1)
+        d2 = data2.reshape(rbs).sum(-1).sum(1)
+        # this makes a pixel masked if any of the summed pixels is masked
+        # if we want to mask only if all the pixels are masked then we could
+        # use prod in place of sum here
+        msk = mask.reshape(rbs).sum(-1).sum(1)
+        # need to convert back to bool after summing
+        msk = msk.astype(bool)
+        # This restores the image to the size of the original images as in the
+        # IDL code:
+        d1tmp = np.zeros_like(data1)
+        d2tmp = np.zeros_like(data2)
+        mtmp = np.zeros_like(mask, dtype=bool)
+        for i in range(binfac):
+            for j in range(binfac):
+                d1tmp[i::binfac, j::binfac] = d1[:]
+                d2tmp[i::binfac, j::binfac] = d2[:]
+                mtmp[i::binfac, j::binfac] = msk[:]
+        data1 = d1tmp
+        data2 = d2tmp
+        mask = mtmp
```

Comment:
Rather than reimplementing rebinning, I would suggest using the [`superpixel` method](https://docs.sunpy.org/en/stable/generated/api/sunpy.map.GenericMap.html#sunpy.map.GenericMap.superpixel) provided on the `Map` class. From a separation of concerns perspective, it may also be better to leave it up to the user to rebin their Map prior to passing them to `xrt_teem`.

Reply:
I've looked into the superpixel method. It seems like a good option. One thing I found annoying about it is that it requires that the dimensions argument have units even though they have to be integers. Is that really necessary (or useful)?
The question of including that in xrt_teem gets to the question of the extent to which we want the XRTpy version to be consistent with the IDL version. Also, people may not need a binned version of thier image, but for the purposes of deriving the temperature and emission measure, the extra signal/noise might be useful. I'm inclined to leave the option to rebin in.

Reply:
> I've looked into the superpixel method. It seems like a good option. One thing I found annoying about it is that it requires that the dimensions argument have units even though they have to be integers. Is that really necessary (or useful)?

In `sunpy`, all inputs and outputs must have units, which includes pixels. If the superpixel method is called within `xrt_teem`, you could always accept a unitless argument to `xrt_teem` and then add `u.pix` when calling `superpixel`. The advantage of using `superpixel` is that `xrtpy` doesn't have to worry about maintaining a custom rebin method that may potentially behave differently than other rebin methods.

> The question of including that in xrt_teem gets to the question of the extent to which we want the XRTpy version to be consistent with the IDL version.

Fair point. I guess I am not aware to what extent the XRT team would like to adhere to the style/interface of the original IDL code. There of course is an argument for wanting to be consistent with what people have used in the past.

> Also, people may not need a binned version of their image, but for the purposes of deriving the temperature and emission measure, the extra signal/noise might be useful. I'm inclined to leave the option to rebin in.

Fair point though I think it is not unreasonable to expect users to do this themselves before passing the map into this function, particularly given that it is doable using a single method on the `Map` object. This also means fewer arguments to the `xrt_teem` function.